sql: don't propagate spans from local processors as metadata

yuzefovich opened the pull request:
**sql: improve how some exec stats are collected**

This commit improves how we collect the scan stats as well as the
contention information. Previously, we would analyze the recording of
a processor that could - in theory - include its children. In practice,
this wasn't the case since at the moment each processor creates
a tracing span with the detached option. However, the follow-up commit
will change that, so this commit clarifies how we do things.

In particular, it introduces a new notion of "barrier" between the
tracing spans, and we now insert a barrier between spans created for
DistSQL processors. The "disconnect" is a lot weaker than the detached
option with the only effect being not including the children behind the
barrier in a newly introduced method
`GetStructuredRecordingUpToBarrier`. This method is now utilized for
both the scan stats and the contention info. (Note that previously for
the latter we were using the "configured" recording which was an
overkill in case the verbose recording is used, and this is now fixed.)
Without this barrier, once the detached option is removed, a processor
could incorrectly think that it experienced contention when, in fact, it
was the processor's input which experienced it. We also cannot simply
look at the structured events of the processor's span while ignoring all
of its children because the contention events are not recorded into that
span directly. This notion of a "barrier" seems like the easiest
solution.

Release note: None

**sql: don't propagate spans from local processors as metadata**

This commit changes how we create tracing spans for components of the
flow on the gateway. Previously, we would always create them with the
detached option, and, as a result, we needed to collect the traces and
propagate them as metadata to be imported by the DistSQLReceiver into
the span of the flow on the gateway. However, this is not necessary to
do so on the gateway - we can just let the regular tracing behavior to
happen there. We only need to be careful about inserting the barrier
(introduced in the previous commit) to be able to attribute the
contention events and the scan stats to the correct processor.

The main benefit of this change is that we now preserve the hierarchy of
the tracing spans in the gateway flows making them much more
understandable when viewing via the jaeger. Perhaps this is also a minor
performance improvement since the native trace collection is probably
faster than having to import the trace data via the metadata, as was
done previously.

An additional improvement can be made to create the spans with the
detached recording only for the root components of the remote flows.
This is left as a TODO for now.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96522)
<!-- Reviewable:end -->

yuzefovich committed:
sql: don't propagate spans from local processors as metadata

This commit changes how we create tracing spans for components of the
flow on the gateway. Previously, we would always create them with the
detached option, and, as a result, we needed to collect the traces and
propagate them as metadata to be imported by the DistSQLReceiver into
the span of the flow on the gateway. However, this is not necessary to
do so on the gateway - we can just let the regular tracing behavior to
happen there. We only need to be careful about inserting the barrier
(introduced in the previous commit) to be able to attribute the
contention events and the scan stats to the correct processor.

The main benefit of this change is that we now preserve the hierarchy of
the tracing spans in the gateway flows making them much more
understandable when viewing via the jaeger. Perhaps this is also a minor
performance improvement since the native trace collection is probably
faster than having to import the trace data via the metadata, as was
done previously.

An additional improvement can be made to create the spans with the
detached recording only for the root components of the remote flows.
This is left as a TODO for now.

Release note: None

yuzefovich committed:
sql: improve how some exec stats are collected

This commit improves how we collect the scan stats as well as the
contention information. Previously, we would analyze the recording of
a processor that could - in theory - include its children. In practice,
this wasn't the case since at the moment each processor creates
a tracing span with the detached option. However, the follow-up commit
will change that, so this commit clarifies how we do things.

In particular, it introduces a new notion of "barrier" between the
tracing spans, and we now insert a barrier between spans created for
DistSQL processors. The "disconnect" is a lot weaker than the detached
option with the only effect being not including the children behind the
barrier in a newly introduced method
`GetStructuredRecordingUpToBarrier`. This method is now utilized for
both the scan stats and the contention info. (Note that previously for
the latter we were using the "configured" recording which was an
overkill in case the verbose recording is used, and this is now fixed.)
Without this barrier, once the detached option is removed, a processor
could incorrectly think that it experienced contention when, in fact, it
was the processor's input which experienced it. We also cannot simply
look at the structured events of the processor's span while ignoring all
of its children because the contention events are not recorded into that
span directly. This notion of a "barrier" seems like the easiest
solution.

Release note: None

andreimatei reviewed (commented):
> Reviewed 16 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96522)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96522#-NNlr3W24UBhPmjWQiuC):*
I don't really like how this barrier has infected a bunch of functions. And I feel like it doesn't really belong in the tracing library to begin with.
Could we burden the two code paths that care about it to deal with filtering out the parts of the trace that they want filtered out, and maybe provide a helper function in sql that doesn't recurse into spans belonging to another processor (by looking at the span name, or some new tag) ?
___
*[`pkg/util/tracing/crdbspan.go` line 1131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96522#-NNlqr_J9884xk75xaM6:-NNlqr_J9884xk75xaM7:b-dnz26i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e4249f1b7f2166b03ecddc2bc1d2cc01b05bb68/pkg/util/tracing/crdbspan.go#L1131)):*
> ```Go
> 			sp.getChildrenMetadataRecursivelyLocked(
> 				childrenMetadata,
> 				// TODO: why is this always true?
> ```

I guess because we want `sp` operation to be reflected in its metadata.
What I don't quite understand is why the other caller to this method passes `false`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96522)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96522#-NNlr3W24UBhPmjWQiuC):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't really like how this barrier has infected a bunch of functions. And I feel like it doesn't really belong in the tracing library to begin with.
Could we burden the two code paths that care about it to deal with filtering out the parts of the trace that they want filtered out, and maybe provide a helper function in sql that doesn't recurse into spans belonging to another processor (by looking at the span name, or some new tag) ?
</blockquote></details>

How can I distinguish children spans between those that I want vs others?

Let's consider a concrete example: imaging we have a query that has an index join on top of a scan, so we have two processors for which we call `execinfra.ProcessorSpan`:
```
spIndexJoin := execinfra.ProcessorSpan("colindexjoin")
spScan := execinfra.ProcessorSpan("colbatchscan")
```
Notably, `spScan` is a child of `spIndexJoin`. Both of these processors will perform reads, and each KV read will get its own span, so overall hierarchy is as follows:
```
      spIndexJoin
     /            \
ijRead            spScan
                     \
                      scanRead
```
When analyzing the trace of `spIndexJoin` for the index join I need to be able to filter out `spScan` and `scanRead` spans (since they will be handled when analyzing `spScan` in the scan processor).

I thought about using tags (via `tracing.Span.SetTag`) where each processor can be uniquely identified by the processor ID. However, it doesn't seem like these tags are inherited by the children.

One option I can see is that if these processor ID tags are added, then we could extend the tracing library to expose a walker for collecting the recording. In particular, that walker would call the provided callback on each child span before recursing into that child, and if the callback says that the span should be ignored (because the callback analyzed the tags and found the ID of another processor), it will be ignored. Thoughts about this idea? Or any other approach?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96522)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96522#-NNlr3W24UBhPmjWQiuC):*
> I thought about using tags (via tracing.Span.SetTag) where each processor can be uniquely identified by the processor ID. However, it doesn't seem like these tags are inherited by the children.

Tags are not inherited (there used to be a type of inherited tag back in the day called "baggage" but we got rid of it cause it was too confusing). But you know how to walk trees and find the closest ancestor with a tag.
But your real problem I think is that you're dealing with the "structured recording", which is not a tree, right? So inherited tags would not help you either.

> we could extend the tracing library to expose a walker for collecting the recording.

I think there's complications with the fact that the recording is collected across both open and child spans, and they look a little different. But yeah, I think this is what we should try - introduce a trace visitor that walks a trace and has the option to stop recursing on a span (based on that span's tags).


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96522)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96522#-NNlr3W24UBhPmjWQiuC):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> I thought about using tags (via tracing.Span.SetTag) where each processor can be uniquely identified by the processor ID. However, it doesn't seem like these tags are inherited by the children.

Tags are not inherited (there used to be a type of inherited tag back in the day called "baggage" but we got rid of it cause it was too confusing). But you know how to walk trees and find the closest ancestor with a tag.
But your real problem I think is that you're dealing with the "structured recording", which is not a tree, right? So inherited tags would not help you either.

> we could extend the tracing library to expose a walker for collecting the recording.

I think there's complications with the fact that the recording is collected across both open and child spans, and they look a little different. But yeah, I think this is what we should try - introduce a trace visitor that walks a trace and has the option to stop recursing on a span (based on that span's tags).
</blockquote></details>

Btw, I would try to make this visitor generic, without tying it to recording collection in any way. Just basically invoke a method for every span, and your visitor will look through the structured events for that and do whatever it wants with them.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96522)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96522#-NNlr3W24UBhPmjWQiuC):*
> But your real problem I think is that you're dealing with the "structured recording", which is not a tree, right?

Yeah, that's problematic. I see that we don't include tags into `RecordedSpan` for structured recording, so how would I go around this? One option is to define "structured tags" which are included into the structured recording (in addition to also being included in verbose recording as all other tags). However, these tags would still not apply to the child spans. So these "structured tags" would also need to be inherited which sounds like you wouldn't support.

The interesting child spans should always be finished by the time I need to examine the parent's trace while the parent itself might or might not be finished (just FYI in case it matters for this discussion).


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Discussed with Andrei offline - will implement this idea via event listeners. Will open a separate PR for that.

yuzefovich commented:
For posterity, superseded by #98831.
